src/drivers/net/mcst/l_e1000.c

```diff
@@ -0,0 +1,728 @@
+	ep->init_block->rdra |= htole32(E1000_LOG_RX_BUFFERS);
+	ep->init_block->tdra = htole32((uint32_t)((uintptr_t)(&ep->tx_ring[0])));
+	ep->init_block->tdra |= htole32(E1000_LOG_TX_BUFFERS);
+	log_debug("Rx Desc Ring DMA Addr 0x%x", le32toh(ep->init_block->rdra));
+	log_debug("Tx Desc Ring DMA Addr 0x%x", le32toh(ep->init_block->tdra));
+	/***********************************************************************/
+	/* low 32 bits */
+	init_block_addr_part = (uint32_t)((uintptr_t)ep->init_block & 0xffffffff);
+	e1000_write_e_base_address(ep, init_block_addr_part);
+	log_debug("Init Block Low  DMA addr: 0x%x", init_block_addr_part);
+	/* high 32 bits */
+	init_block_addr_part = (uint32_t)(((uintptr_t)(ep->init_block) >> 32) & 0xffffffff);
+	e1000_write_dma_base_address(ep, init_block_addr_part);
+	log_debug("Init Block High DMA addr: 0x%x", init_block_addr_part);
+	/************************************************************************/
+
+	return 0;
+}
+
+static int l_e1000_reset(struct net_device *dev) {
+	struct l_e1000_priv *ep = netdev_priv(dev);
+	int i;
+
+	/* Reset the PCNET32 */
+	e1000_write_e_csr(ep, STOP);
+	/* wait for stop */
+	for (i = 0; i < 1000; i++) {
+		if (e1000_read_e_csr(ep) & STOP) {
+			break;
+		}
+	}
+
+	e1000_init_block(dev);
+
+	e1000_write_e_csr(ep, INIT);
+	l_e1000_set_phy_mode(dev);
+
+	sleep(5);
+
+	i = 0;
+	while (i++ < 1000) {
+		if (e1000_read_e_csr(ep) & IDON) {
+			break;
+		}
+	}
+
+	log_debug("e_csr register after initialization: 0x%x, must be 0x%x",
+			e1000_read_e_csr(ep), (IDON | INTR | INIT));
+
+	e1000_write_e_csr(ep, IDON);
+	log_debug("e_csr register after clear IDON bit: 0x%x, must be 0x%x",
+			e1000_read_e_csr(ep), (INIT));
+
+	e1000_write_e_csr(ep, INEA|STRT);
+	log_debug("e_csr register after setting STRT bit: 0x%x, must be 0x%x",
+			e1000_read_e_csr(ep), (INEA | RXON | TXON | STRT | INIT));
+
+	return 0;
+}
+
+static int l_e1000_hw_init(struct pci_slot_dev *pci_dev, struct net_device *dev) {
+	struct l_e1000_priv *ep;
+	struct l_e1000_dma_area *m;
+	unsigned int soft_reset;
+
+	int fdx, mii, gmii;
+	uint32_t val = 0;
+	int i;
+
+	ep = netdev_priv(dev);
+
+	m = &e1000_dma_area;
+	ep->dma_area = m;
+	ep->dma_addr = (uintptr_t)m;
+
+	ep->init_block = &m->init_block;
+	ep->tx_ring = &m->tx_ring[0];
+	ep->rx_ring = &m->rx_ring[0];
+
+	ep->base_ioaddr = dev->base_addr;
+	ep->netdev = dev;
+
+	/* Setup STOP bit; Force e1000 resetting  */
+	e1000_write_e_csr(ep, STOP);
+	/* PHY Resetting */
+	soft_reset = 0;
+	soft_reset |= (E1000_RSET_POLARITY | SRST);
+	e1000_write_mgio_csr(ep, soft_reset); /* startup software reset */
+	soft_reset = e1000_read_mgio_csr(ep);
+	soft_reset &= ~(SRST);
+	e1000_write_mgio_csr(ep, soft_reset); /* stop software reset */
+
+	/* Setup HW (MAC), also known as "Physical" address.
+	 * because unknown now, assumed to be 10:20:30:40:50:60 HEX */
+	for (i = 0; i < 6; i++) {
+		dev->dev_addr[i] = l_base_mac_addr[i];
+	}
```

Comment:
The comment does not correspond to the current `l_base_mac_addr`

Reply:
done